FTS-6 RestController "/api/pitch"  POST endpoint

hulfyback opened the pull request:
I've written this endpoint and I've made 3 tests:
- status code 201
- status code 401
- status code 404

hulfyback committed:
feat(levelup): add pitch api

Create DTO for Pitch
Create /api/pitch endpoint
Create interface for constants

hulfyback committed:
refactor(levelup): remove required body constant

hulfyback committed:
refactor(levelup): add @RequestBody param

In PitchRestControl class add @RequestParam annotation to @PostMapping endpoint method

hulfyback committed:
feat(levelup): add constant to IPitchRestContoller

Add new static final String for required body for using it in tests

hulfyback committed:
test(levelup): test authorization succesful

This test checks if authorization is succesful then response status is 201

hulfyback committed:
test(levelup): test authorization

Test if status is unauthorized then returns status error code 401
Test if there are missing field then returns status error code 404

hulfyback committed:
refactor(levelup): remove default values

Remove default values from model Pitch fields

drexeverest commented on src/main/java/com/greenfoxacademy/levelup/controller/IPitchRestInterface.java:
```diff
@@ -0,0 +1,24 @@
+
+public interface IPitchRestInterface {
+
+  static final String PITCH_REQUIRED_BODY = "{\n"
```
Collection of static strings would be better if placed in a separate class, or well, in a class instead of an interface

drexeverest commented on src/main/java/com/greenfoxacademy/levelup/controller/PitchRestController.java:
```diff
@@ -0,0 +1,39 @@
+  private static boolean isAuthorized = true;
+  private static boolean isMissingField = false;
+
+  public static void setAuthorization(boolean authorized) {
```
This method is not used

drexeverest commented on src/main/java/com/greenfoxacademy/levelup/controller/PitchRestController.java:
```diff
@@ -0,0 +1,39 @@
+    isAuthorized = authorized;
+  }
+
+  public static void setMissingFields(boolean missingField) {
```
It's not used either

drexeverest commented on src/main/java/com/greenfoxacademy/levelup/controller/PitchRestController.java:
```diff
@@ -0,0 +1,39 @@
+
+  @PostMapping("/api/pitch")
+  public ResponseEntity<String> postPitchApi(@RequestBody Pitch pitch) {
+    httpHeaders.set("Authorization", "");
```
You are setting the header by yourself instead of checking the incoming header, which would be the point of it

drexeverest commented on src/main/java/com/greenfoxacademy/levelup/controller/PitchRestController.java:
```diff
@@ -0,0 +1,39 @@
+    if (isAuthorized) {
+      return new ResponseEntity<String>(IPitchRestInterface.PITCH_SUCCESSFUL_BODY, httpHeaders,
+          HttpStatus.CREATED);
+    } else if (isMissingField) {
```
isAuthorized/isMissingField are the values of the boolean variables, please put some logic behind them e.g.
 - check for the header
 - check for the fields

drexeverest commented on src/test/java/com/greenfoxacademy/levelup/controller_test/PitchRestControllerTest.java:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  @Test
+  public void whenStatusIsUnauthorized_thenReturnsStatusCode401() throws Exception {
```
If you have the method to check the header, then you can include an Authorization header here in the test too, so it can be tested like that

drexeverest commented on src/test/java/com/greenfoxacademy/levelup/controller_test/PitchRestControllerTest.java:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  @Test
+  public void whenMissingField_thenReturnsStatusCode404() throws Exception {
```
If your field checker method is done, than you do not nave to set boolean values here, just put the object in the body which has missing fields, and test it like that

lazarsoma committed:
fix(levelup): fixing based on the review

Fixing the model(setters and getters), the controller(changing the
logic, deleting unused methods) and the test(fixing whenMissingField_thenReturnsStatusCode404 test)

lazarsoma committed:
feat(levelup): create hasNullField method

Creating hasNullField method in the model Pitch to check if its field
are empty.

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/controller/PitchRestController.java:
```diff
@@ -12,13 +12,12 @@
 
   @PostMapping("/api/pitch")
-  public ResponseEntity<String> postPitchApi(@RequestHeader(value = "Authorization", required = false) String authorization, @RequestBody(required = false) Pitch pitch) {
+  public ResponseEntity<String> postPitchApi(@RequestHeader(value = "Authorization", required = false) String authorization, @RequestBody(required = false) Pitch pitch) throws IllegalAccessException {
```
ok

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/model/Pitch.java:
```diff
@@ -59,4 +60,13 @@ public void setPitchMessage(String pitchMessage) {
     this.holders = holders;
   }
+
+  public boolean hasNullField() throws IllegalAccessException {
```
ok

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/controller/PitchRestController.java:
```diff
@@ -12,13 +12,12 @@
               HttpStatus.UNAUTHORIZED);
-    } else if (pitch.getBadgeName() == null || pitch.getHolders() == null || pitch.getOldLVL() == 0 ||
-            pitch.getPitchedLVL() == 0 || pitch.getPitchMessage() == null) {
+    } else if (pitch.hasNullField() == true) {
```
ok

rozanagy reviewed (commented):
i am ok with the changes

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/controller/IPitchRestInterface.java:
```diff
@@ -1,24 +1,23 @@
 
-public interface IPitchRestInterface {
+public class IPitchRestInterface {
+  public static String PITCH_REQUIRED_BODY = "{\n"
```
refactor the field name (as not constant anymore)

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/controller/IPitchRestInterface.java:
```diff
@@ -1,24 +1,23 @@
-      + "  \"pitchMessage\": \"Hello World! My English is bloody gorgeous.\",\n"
-      + "  \"holders\": [\"balazs.jozsef\", \"benedek.vamosi\", \"balazs.barna\"]\n"
-      + "}".replaceAll("\\s", "");
+  public static String PITCH_SUCCESSFUL_BODY = "{\n"
```
refactor the field name (as not constant anymore)

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/controller/IPitchRestInterface.java:
```diff
@@ -1,24 +1,23 @@
-  static final String PITCH_SUCCESSFUL_BODY = "{\n"
-      + "  \"message\": \"Created\"\n"
-      + "}".replaceAll("\\s", "");
+  public static String PITCH_UNAUTHORIZED_BODY = "{\n"
```
refactor the field name (as not constant anymore)

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/controller/IPitchRestInterface.java:
```diff
@@ -1,24 +1,23 @@
-  static final String PITCH_UNSUCCESFUL_BODY = "{\n"
-      + "  \"error\": \"Please provide all fields\"\n"
-      + "}".replaceAll("\\s", "");
+  public static String PITCH_UNSUCCESFUL_BODY = "{\n"
```
refactor the field name (as not constant anymore)

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/controller/PitchRestController.java:
```diff
@@ -1,39 +1,28 @@
-    } else if (isMissingField) {
-      return new ResponseEntity<String>(IPitchRestInterface.PITCH_UNSUCCESFUL_BODY, httpHeaders,
-          HttpStatus.NOT_FOUND);
+    if (authorization == null || authorization.equals("")) {
```
ok (already refactored)

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/controller/PitchRestController.java:
```diff
@@ -1,39 +1,28 @@
     }
-    return new ResponseEntity<String>(IPitchRestInterface.PITCH_UNAUTHORIZED_BODY, httpHeaders,
-        HttpStatus.UNAUTHORIZED);
+    return new ResponseEntity<>(IPitchRestInterface.PITCH_SUCCESSFUL_BODY,
```
ok

rozanagy commented on src/main/java/com/greenfoxacademy/levelup/model/Pitch.java:
```diff
@@ -1,11 +1,62 @@
   private List<String> holders;
-}
+
+  public Pitch() {
```
ok

rozanagy commented on src/test/java/com/greenfoxacademy/levelup/controller_test/PitchRestControllerTest.java:
```diff
@@ -33,30 +28,36 @@ public void setup() {
-      .contentType(MediaType.APPLICATION_JSON)
-      .content(IPitchRestInterface.PITCH_REQUIRED_BODY))
-      .andExpect(status().isCreated());
+            .header("Authorization", "author")
```
ok

rozanagy commented on src/test/java/com/greenfoxacademy/levelup/controller_test/PitchRestControllerTest.java:
```diff
@@ -33,30 +28,36 @@ public void setup() {
-      .contentType(MediaType.APPLICATION_JSON)
-      .content(IPitchRestInterface.PITCH_REQUIRED_BODY))
-      .andExpect(status().is4xxClientError());
+            .contentType(MediaType.APPLICATION_JSON)
```
ok

rozanagy commented on src/test/java/com/greenfoxacademy/levelup/controller_test/PitchRestControllerTest.java:
```diff
@@ -33,30 +28,36 @@ public void setup() {
-        .contentType(MediaType.APPLICATION_JSON)
-        .content(IPitchRestInterface.PITCH_REQUIRED_BODY))
-        .andExpect(status().isNotFound());
+    String requiredBody = "{\n"
```
ok

rozanagy reviewed (commented):
please make the necessary changes

lazarsoma committed:
fix(levelup): fix the name of the constants

Changing from uppercase to lowercase.

lazarsoma committed:
fix(levelup):refactor IPitchRestInterface

Creating collection package, moving IPitchRestInterface into the Message
class and deleting IPitchRestinterface

lazarsoma committed:
fix(levelup): fix conflict

Fixing Message class

lazarsoma committed:
merge(levelup): merge zsoltmolnar-FTS-6 into dev
